log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2016-2021 Talsma ICT
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package nl.talsmasoftware.context.log4j2;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.logging.log4j.ThreadContext;
+
+/**
+ * Represents a snapshot of the data of the Log4j 2 {@link ThreadContext} of a
+ * specific thread at a certain point in the past.
+ */
+public class ThreadContextData {
+    private final Map<String, String> contextMap;
+    private final List<String> contextStack;
+
+    /**
+     * Applies the data to the current thread.
+     *
+     * @param data data to apply, may be {@code null}
+     * @param overwrite whether all existing data should overwritten
+     */
+    static void applyToCurrentThread(ThreadContextData data, boolean overwrite) {
```

Comment:
These methods are only intended for usage by `ThreadContextManager`. It might also be possible to have this logic only in `ThreadContextManager`, work with the public methods of `ThreadContextData` and have its constructor be package-private, but I felt the logic belongs closer to `ThreadContextData` than to `ThreadContextManager`.

As for why these methods are package-private: Most of them provide internal functionality already provided (indirectly) through `ThreadContextManager` or through Log4j 2's ThreadContext. Providing it publicly might be confusing and lead to incorrect usage.
Making the whole class package-private is not an option because it is publicly exposed through `ThreadContextManager` and would therefore cause usability issues when someone tries to use `ThreadContextManager`.

I don't think making it an interface fits really well because it is not expected that the user implements or subclasses this class. Because the package-private methods are not accessible to the user and are not shown in the javadoc either, I don't think that this is an issue.

Would it be alright for you to leave the class as is? If you want I can add a comment (not part of the documentation) to the class describing the rationale I described here.